### ISSUE TITLE
left_join with large dataset and multiple matching columns crashes R if adding new rows (cartesian product)

### ISSUE THREAD

**Issue by nickbond** [hadley/dplyr#1230]
The above crash occurred for me on both OS X and windows, but was alleviated by specifying the number of rows in the second table being joined (df2 below had exactly 1130 rows). The first join column was formatted as POSIXct. It is a large dataset and I have not been able to reproduce with a simpler example, but happy to provide the data if you need a reproducible example.

```r
left_join(df1, df2[1:1130,], by = c('date'='date', 'site.2014'='site')).
```

Thanks,
Nick

**Comment by romainfrancois** [comment]
yes please @nickbond provide a reprex.

**Comment by nickbond** [comment]
Dear François,
I've made a copy of the data frames in question, which can be downloaded at the following link (https://www.dropbox.com/s/yw0lyu5dng6r1vb/bond_example.RData?dl=0). 

The code below joins the two dataframes. What I discovered by accident is that including 'zone' in the list of join terms avoids the error (example 1), as does specificying the length of the second data frame (example 2). Example 3 causes a crash. Importantly, 'zone' is not essential (other than maintaining it as a unique field in the joined table) because sites codes are unique within each zone, and thus I would expect the join to work regardless. The main question is why example 2 works, but not example 3. 

Regards, Nick

library(dplyr)
load("bond_example.RData")
left_join(df1, df2, by = c('date'='date', 'zone'='zone', 'site'='site')) #Example 1
left_join(df1, df2[1:1130,], by = c('date'='date', 'site'='site')) #Example 2
left_join(df1, df2, by = c('date'='date', 'site'='site')) #Example 3
